Issue 236: Merge in base Ansible setup steps from legacy playbook deployment

Hzaakk opened the pull request:
# Changes
1. Rewritten file: `bootstrap/ansible/playbook.yml`
2. Playbook now includes production tasks such as:
    a. Firewall setup
    b. Sending logs to LRC server
3. All tasks that report `changed` status despite changing nothing, now report an `ok` status:
    a. `file` module with `touch` state now replaced with `copy` module with empty `content` (file module always reports changed because it changes the last modified time on files)
    b. `Initialize Python virtual environment` task returns a changed state only if `venv/bin/activate` did not exist before. This is done by switching from the `command` module to the `shell` module and using registers and `changed_when`.
4. Playbook is split into different blocks, namely `provisioning`, `development provisioning`, `production provisioning`, and `common`. The provisioning tasks are ideally just run once, aside from when changes affect provisioning. Common tasks are the ones that matter when switching branches and updating the database and similar normal development stuff; they are run more often and are not meant to just be run once. Some tasks (especially production) are dependent on BRC vs LRC and use the `when` argument with the flags as necessary.
5. Two new boolean variables in `main.yml`: `production_tasks`, and `common_tasks`, set which of the corresponding tasks are run by default when you run the playbook without arguments. Both default to true (which is needed initially when the VM is made)
6. Tasks are also tagged `production` or `common`, so one can use `ansible-playbook bootstrap/ansible/playbook.yml --tags common` or `--tags production` or `--skip-tags common` to skip redundant tasks and lower playbook execution time.
7. Vagrantfile now points to `ansible/playbook.yml` instead of `development/playbook.yml`
## TODO:
1. Maybe reorganize files in `bootstrap/ansible` and `bootstrap/development`
2. Add firewall rules for LRC in the LRC production tasks.
3. Test production tasks (they were copied from the old playbook Matthew gave me).
# How I tested
1. I destroyed my VM and reran `vagrant up`, noticing no problems with basic website navigation.
2. I then switched branches and ran `ansible-playbook bootstrap/development/playbook.yml --tags common` to run the new migrations and had no problems with basic website navigation.
# Notes
All the previous tasks in `development/playbook.yml` are included, but some were switched around to be moved to the common tasks. I don't know if that could cause any issues. I duplicated the Postgres restarting task just in case.

matthew-li commented on bootstrap/ansible/playbook.yml:
```diff
@@ -1,284 +1,491 @@
+          become_user: "{{ djangooperator }}"
+          changed_when: venv_register.stdout == "created"
+
+        # Install Django application dependencies, so that psycopg2, which is
```
We could make this common, since we do add requirements. It certainly isn't true to say that this "can be skipped always after the first time".

matthew-li commented on bootstrap/ansible/playbook.yml:
```diff
@@ -1,284 +1,491 @@
+          shell: "{{ git_prefix }}/venv/bin/mod_wsgi-express install-module --modules-directory /etc/httpd/modules > /etc/httpd/conf.modules.d/02-wsgi.conf"
+          when: not mod_wsgi_installed.stat.exists
+
+        - name: Create a WSGI configuration file from a template if SSL is disabled
```
We may also make changes to the Apache configuration files.

matthew-li commented on bootstrap/ansible/playbook.yml:
```diff
@@ -1,284 +1,491 @@
+
+        # Configure the Django application.
+
+        - name: Copy the ColdFront local_settings.py file from the sample
```
Settings templates change very often, so these should be refreshed each time. (Everything including and below the WSGI configuration file generation should be moved to `common`.)

matthew-li commented on bootstrap/ansible/playbook.yml:
```diff
@@ -1,284 +1,491 @@
+                destfile: /etc/rsyslog.conf
+                insertbefore: "# ### begin forwarding rule ###"
+                block: |
+                  # Send copyof logs to LBL's central syslog server
```
"copy of"

matthew-li commented on bootstrap/ansible/playbook.yml:
```diff
@@ -1,284 +1,491 @@
+
+          when: flag_lrc_enabled
+
+      when: provisioning_tasks == true and deployment == "prod"
```
There is also a "staging" deployment.

Hzaakk replied:
I just changed the conditional to `... and (deployment == "prod" or deployment == "staging")` is that fine or should staging deployments not send logs to LBL servers, and anything else?

matthew-li replied:
Can we have a third task: "Staging only provisioning tasks" that we leave blank for now?

The inconvenience at the moment is that BRC staging is running on an LBL VM (the same VM as LRC staging)--I'd like to set up a different staging server at some point.

matthew-li reviewed (changes requested):
Thanks for working on this. I've noted some tasks that should be moved around, and I'll test on a new box afterward.

Some other things:
- Can you update `README.md` to note changes to the VM provisioning process, and anything someone should know when deploying to production?
- Will we be able to retire `playbook.yml` and `main.copyme` in `development`? (You did note we should move around files in `development`, so I'm guessing yes. We may as well do that now.)

matthew-li commented on Vagrantfile:
```diff
@@ -12,7 +12,7 @@ Vagrant.configure("2") do |config|
   config.vm.provision "ansible_local" do |ansible|
-    ansible.playbook = "coldfront_app/coldfront/bootstrap/development/playbook.yml"
+    ansible.playbook = "coldfront_app/coldfront/bootstrap/ansible/playbook.yml"
     ansible.galaxy_role_file = "coldfront_app/coldfront/bootstrap/development/requirements.yml"
```
The requirements file is accounted for when provisioning the dev. VM.

On production, in the past, I installed requirements manually as needed (e.g., `ansible-galaxy collection install community.general`). Is there a command to install them from the `.yml`, and if so, can you include it in `README.md`?

Hzaakk commented:
> * Will we be able to retire `playbook.yml` and `main.copyme` in `development`? (You did note we should move around files in `development`, so I'm guessing yes. We may as well do that now.)

I'm not sure about the `main.copyme` in `development` because it looks like the one in `ansible` is filled in for production and the one in `development` is filled in for development (with differences like port numbers, file paths, etc).

The `playbook.yml` in `development` should have identical functionality to the new one in `ansible` (assuming changing the task order didn't affect anything) if the `deployment` variable is set to `dev`, and both provisioning and common tasks are run. I guess this means it can be retired.

Hzaakk commented:
> * Can you update `README.md` to note changes to the VM provisioning process, and anything someone should know when deploying to production?
> * On production, in the past, I installed requirements manually as needed (e.g., ansible-galaxy collection install community.general). Is there a command to install them from the .yml, and if so, can you include it in README.md?

@matthew-li I've included using `ansible-galaxy collection install -r bootstrap/development/requirements.yml` in the README along with some other changes. Could you review them?

Hzaakk commented:
I also noticed that the Django management commands never return a changed status, even when initially provisioning, maybe there's a way to change that somehow?

![image](https://user-images.githubusercontent.com/43729496/183263763-7211ebd9-5ca2-4a53-9b70-1b9b65696014.png)

matthew-li commented on bootstrap/ansible/playbook.yml:
```diff
@@ -1,293 +1,513 @@
+            state: restarted
+            enabled: yes
+
+        - name: Restart and enable PostgreSQL service
```
I wonder if we need to restart this each time?

Hzaakk replied:
I wasn't sure if later tasks depended on using the restarted database so I included it after the initial postgres tasks just in case. I can remove the task if there arent any dependencies. Depending on the dependencies of the other restarting tasks, we may also be able to use Ansible "handlers", which only run when specified tasks return a "changed" output. Handlers can remove unnecessary restarting and speed up playbook execution.

matthew-li replied:
I don't think anything under "Run Common Tasks" requires that PostgreSQL be restarted first.

Can we perform the "enable" during provisioning?

Hzaakk replied:
When provisioning is enabled, the common tasks are run (the conditional is `when: provisioning_tasks == true or common_tasks == true`) so enabling happens under provisioning. I'll remove the first restart.

matthew-li commented on bootstrap/ansible/playbook.yml:
```diff
@@ -216,6 +216,18 @@
           become_user: "{{ djangooperator }}"
           changed_when: venv_register.stdout == "created"
 
+        # Install psycopg2 and mod-wsgi, which are required to create the
```
I'd prefer not to have the specific modules and versions duplicated here. Given that this is intended to run once, can we install all Pip requirements here?

We'd still need to install them in common tasks, as requirements may change.

matthew-li commented:
> > * Will we be able to retire `playbook.yml` and `main.copyme` in `development`? (You did note we should move around files in `development`, so I'm guessing yes. We may as well do that now.)
> 
> @matthew-li I'm not sure about the `main.copyme` in `development` because it looks like the one in `ansible` is filled in for production and the one in `development` is filled in for development (with differences like port numbers, file paths, etc).
> 
> The `playbook.yml` in `development` should have identical functionality to the new one in `ansible` (assuming changing the task order didn't affect anything) if the `deployment` variable is set to `dev`, and both provisioning and common tasks are run. I guess this means it can be retired.

Right, thanks for the reminder about the default values in `main.copyme`. We could consolidate them by adding a `dev_settings` section to the one in the `ansible` directory. The only other difference in the "BRC/LRC Config" section from a glance are the "Email settings", which could be moved down to the individual settings.

matthew-li commented:
I'm going to try to provision a new VM using the playbook and see if anything is amiss. After that, we should be able to retire `development/playbook.yml`, and then I think this can be merged.

The later test will be to use this to provision the LRC production machine (soon, TBD). That will likely involve some updates to "LRC-only Production Tasks".

matthew-li commented:
@Hzaakk The VM provisioned successfully, and running the playbook again without provisioning tasks works nicely (seems to be faster as well). Thanks!

There are just some lingering comments from above remaining.

Hzaakk commented:
> Right, thanks for the reminder about the default values in `main.copyme`. We could consolidate them by adding a `dev_settings` section to the one in the `ansible` directory. The only other difference in the "BRC/LRC Config" section from a glance are the "Email settings", which could be moved down to the individual settings.
> 
> EDIT: If doing this, the directory in step 6 in the README should be updated.

What would a `dev_settings` section look like? Would it just be moving `development/main.copyme` to `ansible/dev_settings.copyme` or would it be adding comments to `ansible/main.copyme` to indicate needed changes for development? Or would it be something else?

Hzaakk commented:
Along with different emails, the ansible/main.copyme uses email port 25 instead of the 1025 used in development/main.copyme. Is it fine to just keep it at 25, or should I add a TODO to change it to 1025 for development?

Hzaakk commented:
@matthew-li I've addressed every comment I think. I've also moved `development/requirements.yml` to `ansible/requirements.yml` so that everything related to ansible is all consolidated within the one directory.

matthew-li commented:
> Along with different emails, the ansible/main.copyme uses email port 25 instead of the 1025 used in development/main.copyme. Is it fine to just keep it at 25, or should I add a TODO to change it to 1025 for development?

It should be 1025 for development, in line with the debugging server mentioned [here](https://github.com/ucb-rit/coldfront#emails). Looks like you've already included it.
